core: FS: make dirfile interface accept empty object ID

jforissier opened the pull request:
A fix to allow persistent objects with an empty ID.

Sadiq1811 commented:
Reported-by: Sadiq Hussain <sadiq.muchumarri@intel.com>

jforissier commented on core/tee/fs_dirfile.c:
```diff
@@ -198,26 +207,17 @@ TEE_Result tee_fs_dirfile_find(struct tee_fs_dirfile_dirh *dirh,
-	struct dirfile_entry dent;
+	TEE_Result res = TEE_SUCCESS;
+	struct dirfile_entry dent = { };
 	int n;
```
Yes

jforissier commented on core/tee/fs_dirfile.c:
```diff
@@ -267,11 +297,13 @@ TEE_Result tee_fs_dirfile_rename(struct tee_fs_dirfile_dirh *dirh,
 	memset(&dent, 0, sizeof(dent));
 	dent.uuid = *uuid;
 	memcpy(dent.oid, oid, oidlen);
+	if (!oidlen)
```
Ack

jforissier commented on core/tee/fs_dirfile.c:
```diff
@@ -198,26 +207,17 @@ TEE_Result tee_fs_dirfile_find(struct tee_fs_dirfile_dirh *dirh,
+			continue;
 		if (dent.oidlen != oidlen)
 			continue;
 
```
I agree, will fix

jforissier commented:
Update: comments addressed and NULL check removed in `libutee`.

etienne-lms commented:
`Reviewed-by: Etienne Carriere <etienne.carriere@linaro.org>` for commits
"core: FS: make dirfile interface accept empty object ID".
and "libutee: TEE_{Create,Open}PersistentObject() should accept NULL object ID".

`Acked-by: Etienne Carriere <etienne.carriere@linaro.org>`
for commit "core: mm: vm_check_access_rights(): return success when len == 0"
since I did went though possible side effects in the implementation.

jenswi-linaro commented on core/mm/vm.c:
```diff
@@ -1240,6 +1240,9 @@ TEE_Result vm_check_access_rights(const struct user_mode_ctx *uctx,
 	size_t addr_incr = MIN(CORE_MMU_USER_CODE_SIZE,
 			       CORE_MMU_USER_PARAM_SIZE);
 
+	if (!len)
```
We aren't as forgiving in `core_is_buffer_inside()` and friends.

Prior to this change a successful result from this function guaranteed that buffer indeed was supplied. If we are to change this we need to check all the call sites to see that we don't depend on the old behaviour.

jenswi-linaro reviewed (commented):
Comment for the commit "core: mm: vm_check_access_rights(): return success when len == 0"

jenswi-linaro commented on core/tee/fs_dirfile.c:
```diff
@@ -27,6 +27,16 @@ struct dirfile_entry {
 
+/*
+ * An object can have an ID of size zero. This anonymous object is represented
+ * by oidlen == 0 and oid[0] != 0. When both are zero, the entry is not a valid
```
Why should anything but a specific value written earlier be accepted?

jenswi-linaro commented on core/tee/fs_dirfile.c:
```diff
@@ -27,6 +27,16 @@ struct dirfile_entry {
+ * by oidlen == 0 and oid[0] != 0. When both are zero, the entry is not a valid
+ * object.
+ */
+static inline bool is_free(struct dirfile_entry *dent)
```
"inline" isn't needed.

jforissier replied:
Right

jenswi-linaro commented on core/tee/fs_dirfile.c:
```diff
@@ -238,6 +238,36 @@ TEE_Result tee_fs_dirfile_find(struct tee_fs_dirfile_dirh *dirh,
+	}
+
+	dfh->idx = n;
+	dfh->file_number = dent.file_number;
```
"file_number" and "hash" is always ignored by the caller and always invalid since they are from a free entry.

jenswi-linaro reviewed (commented):
Comments for the commit "core: FS: make dirfile interface accept empty object ID"

jenswi-linaro reviewed (commented):
For the commit "libutee: TEE_{Create,Open}PersistentObject() should accept NULL object ID"
`Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>`

jforissier commented on core/tee/fs_dirfile.c:
```diff
@@ -27,6 +27,16 @@ struct dirfile_entry {
 };
 
+/*
+ * An object can have an ID of size zero. This anonymous object is represented
```
I see what you mean. I'll change.

jforissier commented on core/tee/fs_dirfile.c:
```diff
@@ -267,11 +297,19 @@ TEE_Result tee_fs_dirfile_rename(struct tee_fs_dirfile_dirh *dirh,
+		 * Null sized OID needs a non-null value to differentiate from
+		 * free OIDs. See is_free().
+		 */
+		dent.oid[0] = 1;
```
I will introduce a define.

jforissier commented on core/tee/fs_dirfile.c:
```diff
@@ -238,6 +238,36 @@ TEE_Result tee_fs_dirfile_find(struct tee_fs_dirfile_dirh *dirh,
 	return TEE_SUCCESS;
 }
 
+static TEE_Result tee_fs_dirfile_new(struct tee_fs_dirfile_dirh *dirh,
```
Will do.

jenswi-linaro reviewed (commented):
For commit "core: storage: do not check object ID buffer when its size is zero"
`Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>`

jenswi-linaro commented on core/tee/fs_dirfile.c:
```diff
@@ -238,33 +242,24 @@ TEE_Result tee_fs_dirfile_find(struct tee_fs_dirfile_dirh *dirh,
 
-static TEE_Result tee_fs_dirfile_new(struct tee_fs_dirfile_dirh *dirh,
-				     struct tee_fs_dirfile_fileh *dfh)
+static TEE_Result tee_fs_dirfile_find_empty_idx(struct tee_fs_dirfile_dirh *dh,
```
minor: A new static function doesn't usually need a long prefix, `find_empty_idx()` could have been enough.

jenswi-linaro reviewed (commented):
For "core: FS: make dirfile interface accept empty object ID"
`Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>`

jforissier commented:
Squashed and rebased, thanks @jenswi-linaro.

@etienne-lms I have not kept your R-b for "core: FS: make dirfile interface accept empty object ID" since the commit has changed significantly.
